src/crimson/os/seastore/object_data_handler.cc

```diff
@@ -1083,308 +116,634 @@ ObjectDataHandler::write_ret ObjectDataHandler::prepare_data_reservation(
-                      &pin, &object_data, &to_write](auto maybe_indirect_extent) {
-              auto read_bl = maybe_indirect_extent.get_bl();
-              ceph::bufferlist write_bl;
-              write_bl.substr_of(read_bl, 0, size - pin_offset);
-              write_bl.append_zero(append_len);
-              LOG_PREFIX(ObjectDataHandler::trim_data_reservation);
-              TRACET("First pin overlaps the boundary and has unaligned data"
-                "create data at addr:{}, len:0x{:x}",
-                ctx.t, pin.get_key(), write_bl.length());
-	      to_write.push_back(extent_to_write_t::create_data(
-	        pin.get_key(),
-	        write_bl));
-	      to_write.push_back(extent_to_write_t::create_zero(
-	        (object_data.get_reserved_data_base() + roundup_size).checked_to_laddr(),
-	        object_data.get_reserved_data_len() - roundup_size));
-              return clear_iertr::now();
-            });
-          }
-	}
-      }).si_then([ctx, size, &to_write, &object_data, &pins, this] {
-        return seastar::do_with(
-          prepare_ops_list(pins, to_write,
-	    delta_based_overwrite_max_extent_size),
-          [ctx, size, &object_data](auto &ops) {
-            return do_remappings(ctx, ops.to_remap
-            ).si_then([ctx, &ops] {
-              return do_removals(ctx, ops.to_remove);
-            }).si_then([ctx, &ops] {
-              return do_insertions(ctx, ops.to_insert);
-            }).si_then([size, &object_data] {
-	      if (size == 0) {
-	        object_data.clear();
-	      }
-	      return ObjectDataHandler::clear_iertr::now();
-            });
-        });
-      });
+  if (mapping.get_val().is_zero()) {
+    bufferlist bl;
+    bl.append_zero(len);
+    return ObjectDataHandler::read_iertr::make_ready_future<
+      bufferlist>(std::move(bl));
+  } else {
+    return ctx.tm.read_pin<ObjectDataBlock>(ctx.t, mapping
+    ).si_then([offset, len](auto maybe_indirect_left_extent) {
+      auto read_bl = maybe_indirect_left_extent.get_bl();
+      ceph::bufferlist prepend_bl;
+      prepend_bl.substr_of(read_bl, offset, len);
+      return ObjectDataHandler::read_iertr::make_ready_future<
+	bufferlist>(std::move(prepend_bl));
     });
+  }
 }
 
-/**
- * get_to_writes_with_zero_buffer
- *
- * Returns extent_to_write_t's reflecting a zero region extending
- * from offset~len with headbl optionally on the left and tailbl
- * optionally on the right.
- */
-extent_to_write_list_t get_to_writes_with_zero_buffer(
-  laddr_t data_base,
-  const extent_len_t block_size,
-  objaddr_t offset, extent_len_t len,
-  std::optional<ceph::bufferlist> &&headbl,
-  std::optional<ceph::bufferlist> &&tailbl)
-{
-  auto zero_left = p2roundup(offset, (objaddr_t)block_size);
-  auto zero_right = p2align(offset + len, (objaddr_t)block_size);
-  auto left = headbl ? (offset - headbl->length()) : offset;
-  auto right = tailbl ?
-    (offset + len + tailbl->length()) :
-    (offset + len);
+struct overwrite_params_t {
```

Comment:
Pass it among many methods, and arbitrarily modify some members in the middle is not easy to follow.

Ideally, for such struct with complex states, maintain and update only in high-level logic, and isolate it from self-contained lower-level methods with clear boundary.